src/AJAX.php

```diff
@@ -49,131 +50,83 @@ public function __construct( $jwt, $user ) {
-			'shipping_postcode'      => $billing_address['postal_code'],
-			'shipping_address_1'     => $billing_address['street_address'],
-			'shipping_address_2'     => $billing_address['street_address_2'],
-			'shipping_phone'         => $id_token['phone'],
-			'shipping_email'         => $userdata['user_email'],
-			User::$refresh_token_key => $refresh_token,
-		);
-
-		// Remove empty fields (based on default value).
-		$userdata['meta_input'] = array_filter(
-			$userdata['meta_input'],
-			function ( $field ) {
-				return ! empty( $field );
-			}
-		);
-
-		// If the user is already logged in, save the refresh token (to be used for klarna_access_token), and return.
-		// Otherwise, their cart will be emptied when changing account.
-		$user_id = get_current_user_id();
-		$guest   = 0;
-		if ( $guest !== $user_id ) {
-			$this->user->set_access_token( $user_id, $jwt_access_token, $expires_in );
-			update_user_meta( $user_id, User::$refresh_token_key, $refresh_token );
-
-			wp_send_json_success( 'user already logged in' );
-		}
+		$id_token = $this->jwt->get_payload( $tokens['id_token'] );
+		$userdata = $this->user->get_user_data( $id_token, $refresh_token );
 
 		if ( username_exists( $userdata['user_login'] ) || email_exists( $userdata['user_email'] ) ) {
-			$user = get_user_by( 'login', $userdata['user_login'] );
-			$user = ! empty( $user ) ? $user : get_user_by( 'email', $userdata['user_email'] );
-			if ( empty( $user ) ) {
-				wp_send_json_error( 'user exists, failed to retrieve user data' );
-			}
-
-			// Skip if the user is already logged in.
-			if ( get_current_user_id() === $user->ID ) {
-				wp_send_json_success( 'user exists, already logged in' );
-			}
-
-			// Try to log the user in. The client should refresh the page.
-			$this->user->set_current_user( $user->ID );
-			$this->user->set_access_token( $user->ID, $jwt_access_token, $expires_in );
-
-			wp_send_json_success( 'user exists, logging in' );
+			$user_id = $this->user->merge_with_existing_user( $userdata );
+		} else {
+			$user_id = $this->user->register_new_user( $userdata );
 		}
 
-		$user_id = wp_insert_user( apply_filters( 'siwk_create_new_user', $userdata ) );
 		if ( is_wp_error( $user_id ) ) {
-			wp_send_json_error( 'could not create user' );
+			wp_send_json_error( $user_id->get_error_message() );
+		}
+
+		$guest        = 0;
+		$current_user = get_current_user_id();
+		if ( $guest === $current_user ) {
+			$this->user->sign_in_user( $user_id, $tokens, $refresh_token );
+		} else {
+			// The only condition for displaying the sign-in button is that the user does not have a refresh token.
+			// Therefore, it could be displayed for a signed-in user. If the user is already signed in, we only update the tokens.
+			$this->user->set_tokens( $user_id, $tokens, $refresh_token );
 		}
 
-		do_action( 'woocommerce_created_customer', $user_id, $userdata, false );
+		// phpcs:ignore -- Nonce is checked by calling function.
+		$url  = wc_get_var( $_POST['url'] );
```

Comment:
Should this not be sanitized and unescaped properly? Right now we use without any sanitation.